Register user-defined-network (UDN) binding on kubevirt CR

RamLavi opened the pull request:
**What this PR does / why we need it**:
This PR adds support to user-defined-network binding to the kubevirt CR. 
This is part of the process described on [kubevirt user guide](https://kubevirt.io/user-guide/network/network_binding_plugins/#definition-flow).
The behavior added is currently dev-preview, and as such is protected behind a FG: `PrimaryUserDefinedNetworkBinding`, that is set to `false` by default.

**Reviewer Checklist**
<!-- Check [Expectations from a PR](/CONTRIBUTING.md#expectations-from-a-pr) for the details -->

> Reviewers are supposed to review the PR for every aspect below one by one. To check an item means the PR is either "OK" or "Not Applicable" in terms of that item. All items are supposed to be checked before merging a PR. 

- [x] PR Message
- [x] Commit Messages
- [x] How to test
- [x] Unit Tests
- [x] Functional Tests
- [x] User Documentation
- [x] Developer Documentation
- [ ] Upgrade Scenario
- [ ] Uninstallation Scenario
- [ ] Backward Compatibility
- [ ] Troubleshooting Friendly

**Jira Ticket**:
<!--  Write the link to the Jira ticket:
If the task is not tracked by a Jira ticket, just write "NONE".
-->
```jira-ticket

```

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
NONE
```

nunnatsa commented on controllers/operands/kubevirt.go:
```diff
@@ -45,6 +46,12 @@ const (
+const (
+	// Needs to align with the NAD that will be deployed by CNAO
+	primaryUDNNetworkBindingName      = "primaryUserDefinedNetwork"
+	primaryUDNNetworkBindingNamespace = "default"
```
why in the default ns?

RamLavi replied:
In order for the NetworkAttachmentDefinition to be available to all VMs on all namespaces, it needs to be deployed on `default` namespace.
For the dev-preview we think this is a good enough approach, but IMO in the future we will need to think of other ways.

Also note that HCO I changed the PR since when we talked and not it is not not deploying the NetworkAttachmentDefinition (will happen on CNAO)

nunnatsa commented on controllers/operands/kubevirt.go:
```diff
@@ -421,6 +428,19 @@ func getKVConfig(hc *hcov1beta1.HyperConverged) (*kubevirtcorev1.KubeVirtConfigu
 	seccompConfig := getKVSeccompConfig()
 
+	if hc.Spec.FeatureGates.PrimaryUserDefinedNetworkBinding != nil && *hc.Spec.FeatureGates.PrimaryUserDefinedNetworkBinding {
+		if hc.Spec.NetworkBinding == nil {
+			hc.Spec.NetworkBinding = make(map[string]kubevirtcorev1.InterfaceBindingPlugin)
+		}
+
+		var sidecarImage string
+		var ok bool
+		if sidecarImage, ok = os.LookupEnv(hcoutil.PrimaryUDNImageEnvV); !ok {
+			return nil, errors.New("failed to get primary UDN image env Var")
+		}
+		hc.Spec.NetworkBinding[primaryUDNNetworkBindingName] = primaryUserDefinedNetworkBinding(sidecarImage)
```
Why do we need to modify the HyperConverged CR? we'll need really good reason for that. the spec is for the user.

RamLavi replied:
We can't directly change the kubevirt CR. If we do, HCO will reconcile back to what's on hc.Spec.NetworkBinding..

nunnatsa replied:
HCO currently just copies the `NetworkBinding` map to KV. I think we can change this code with the new logic.

e.g., instead of 
```go
		NetworkConfiguration: &kubevirtcorev1.NetworkConfiguration{
			NetworkInterface: string(kubevirtcorev1.MasqueradeInterface),
			Binding:          hc.Spec.NetworkBinding,
		},
```
do something like 
```go
		NetworkConfiguration: &kubevirtcorev1.NetworkConfiguration{
			NetworkInterface: string(kubevirtcorev1.MasqueradeInterface),
			Binding:          getNetworkBinding(hc),
		},
```

nunnatsa replied:
and the new logic will get into the new function `getNetworkBinding`

RamLavi replied:
DONE

nunnatsa commented on deploy/images.csv:
```diff
@@ -34,3 +34,4 @@ KUBEVIRT_CONSOLE_PROXY_IMAGE,quay.io/kubevirt-ui/kubevirt-apiserver-proxy,KUBEVI
 AAQ_OPERATOR_IMAGE,quay.io/kubevirt/aaq-operator,AAQ_VERSION,002513b78dfd795f5195534b651eace136ead3d73c8ac8cae145607e79c8a8ba
 AAQ_SERVER_IMAGE,quay.io/kubevirt/aaq-server,AAQ_VERSION,9db6ca9147b1d75ba4d6bb1326c4ec289327eb9f8aadbcd951fcc689e584cd0d
 AAQ_CONTROLLER_IMAGE,quay.io/kubevirt/aaq-controller,AAQ_VERSION,170abd7a75397ddce07ba0d64886ead25183c49cacbccf0d191a4dd37ca013fd
+PRIMARY_USER_DEFINED_NETWORK_BINDING_IMAGE,quay.io/kubevirt/network-passt-binding,KUBEVIRT_VERSION,3fb5f9fb87a9ab1f06bed2b2358ad862ef66bc47dc73044f1076a3eb9a4c55ff
```
please move it up, to be grouped with all the lines that use the `KUBEVIRT_VERSION` variable.

RamLavi replied:
DONE

nunnatsa commented on hack/build-manifests.sh:
```diff
@@ -325,7 +326,8 @@ ${PROJECT_ROOT}/tools/manifest-templator/manifest-templator \
   --webhook-image="${HCO_WEBHOOK_IMAGE}" \
-  --cli-downloads-image="${HCO_DOWNLOADS_IMAGE}"
+  --cli-downloads-image="${HCO_DOWNLOADS_IMAGE}" \
+  --primary-udn-image="${}"
```
you need use the `PRIMARY_USER_DEFINED_NETWORK_BINDING_IMAGE` var here

nunnatsa replied:
Also, the parameter name is wrong: should be `primary-udn-binding-image`

RamLavi replied:
oh damn, I though I got everything.
Thanks for the catch!

nunnatsa commented on api/v1beta1/hyperconverged_types.go:
```diff
@@ -507,6 +507,14 @@ type HyperConvergedFeatureGates struct {
+	// Note: this feature is in Developer Preview.
+	// +optional
+	// +kubebuilder:default=false
+	// +default=false
```
Please update the default values also for the `featureGates` field in the `HyperConvergedSpec` struct, and the `spec` field in the `HyperConverged` struct.

RamLavi replied:
Not sure what you mean there.. Can you elaborate?

nunnatsa replied:
the default values appears also in the places I mentioned above.

RamLavi replied:
DONE

nunnatsa reviewed (commented):
Thanks for the PR. Added some inline comments.

My main comment is about modifying the HC CR in kubevirt.go. We need to avoid that. 

also, some TODOs, yes, I know it's WIP...
1. please update docs/cluster-configuration.md
2. please update the "feature gate defaults" test in tests/func-tests/defaults_test.go

RamLavi commented:
[Change](https://github.com/kubevirt/hyperconverged-cluster-operator/compare/7e5a7840b622b7cee7afc5f44f3b404afa3c96cf..a34a83a7a7c62b1d1f3892c9c8dccffbe9667eaa): Added unit tests.

RamLavi commented:
[Change](https://github.com/kubevirt/hyperconverged-cluster-operator/compare/a34a83a7a7c62b1d1f3892c9c8dccffbe9667eaa..fb4397b5987690af88ca7c2ab88431444e8db0c3): Added kubebuilder default labels on HCO CR tree, Added documentation for new FG

RamLavi commented:
[Change](https://github.com/kubevirt/hyperconverged-cluster-operator/compare/fb4397b5987690af88ca7c2ab88431444e8db0c3..bb5b1b06384239baaa9e5ac0b268a12c42dda693): auto-generating manifests, rebase

nunnatsa commented on controllers/operands/kubevirt.go:
```diff
@@ -736,6 +768,21 @@ func getKVDevConfig(hc *hcov1beta1.HyperConverged) *kubevirtcorev1.DeveloperConf
+		},
+		ComputeResourceOverhead: &corev1.ResourceRequirements{
+			Requests: corev1.ResourceList{
+				corev1.ResourceMemory: resource.MustParse("500Mi"),
```
This magic number of 500Mi is not clear. Can we use a constant with meaningful name, that will self explain it?

RamLavi replied:
DONE.

nunnatsa commented on controllers/operands/kubevirt.go:
```diff
@@ -509,10 +521,30 @@ func getKVConfig(hc *hcov1beta1.HyperConverged) (*kubevirtcorev1.KubeVirtConfigu
+		var sidecarImage string
+		var ok bool
+		if sidecarImage, ok = os.LookupEnv(hcoutil.PrimaryUDNImageEnvV); !ok {
+			return nil, errors.New("failed to get primary UDN image env Var")
```
Not sure about this error. This is not something the user can fix. Maybe it will better to check it on boot time and exit, to fail the pod? @tiraboschi WDYT?

tiraboschi replied:
I agree: this is something we are configuring at the bundle build time as part of the opinionated deployment and it's not something the user can or should try to fix.

nunnatsa commented on docs/cluster-configuration.md:
```diff
@@ -249,6 +249,14 @@ See [below](#configure-application-aware-quota-aaq) for Application Aware Quota
+### primaryUserDefinedNetworkBinding Feature Gate
+Set the `primaryUserDefinedNetworkBinding` feature gate to deploy the needed configurations for kubevirt users, so they 
+can bind their VM to a user-defined network (UDN) on the VM's primary interface.
+
```
do we have some kind of documentation we can add as a reference? I'm afraid this description is a bit crypt.

RamLavi replied:
@maiqueb do we?

maiqueb replied:
what documentation are you hoping for @nunnatsa ? 

I'm afraid all we have is the [openshift enhancement document](https://github.com/openshift/enhancements/pull/1623).

oshoval replied:
maybe worth to add the env var that can be set in order to determine non default value
(this what we are doing for example on CNAO readme)
together with the current info, it can give a simple user guide (please consider adding few words that will make it user guide in case it miss)

RamLavi replied:
> maybe worth to add the env var that can be set in order to determine non default value (this what we are doing for example on CNAO readme) together with the current info, it can give a simple user guide (please consider adding few words that will make it user guide in case it miss)

Not sure what you mean here. what env var? This FG is set by HCO CR, and is set to `false` by default.. In order to activate it, no env var is required - only changing the FG value to `true`. 

> also good to have please on PR desc the important details, such as the feature gate name and so on

DONE

RamLavi replied:
> I'm afraid all we have is the [openshift enhancement document](https://github.com/openshift/enhancements/pull/1623).

@nunnatsa will that be sufficient?

oshoval replied:
> Not sure what you mean here. what env var? This FG is set by HCO CR, and is set to `false` by default.. In order to activate it, no env var is required - only changing the FG value to `true`.
> 

I meant the one that allows to override the image, but no biggie

RamLavi replied:
ah.. that's not something a user should be exposed to imo.

RamLavi commented:
[Change](https://github.com/kubevirt/hyperconverged-cluster-operator/compare/bb5b1b06384239baaa9e5ac0b268a12c42dda693..a541cb04ada808fd5a288be9c0597f29058b810b): Add meaningful name const.

RamLavi commented:
[Change](https://github.com/kubevirt/hyperconverged-cluster-operator/compare/a541cb04ada808fd5a288be9c0597f29058b810b..d9e4828c16cbbabe999b4e54a2b8fd89b89d2b3a): fix Sanity issue

RamLavi commented:
[Change](https://github.com/kubevirt/hyperconverged-cluster-operator/compare/d9e4828c16cbbabe999b4e54a2b8fd89b89d2b3a..04a9d2728b5f90d39c6d1b37baa15bfc6878c78a): Fix defaults e2e test

RamLavi commented:
[Change](https://github.com/kubevirt/hyperconverged-cluster-operator/compare/04a9d2728b5f90d39c6d1b37baa15bfc6878c78a..cdc96075d4f2ca648aa6659f8171e10f4a98ac44): change hco-deployment flag to confirm with other image flags format

maiqueb commented:
cc @oshoval

maiqueb commented:
/cc @oshoval

oshoval commented on hack/build-manifests.sh:
```diff
@@ -372,7 +373,8 @@ ${PROJECT_ROOT}/tools/csv-merger/csv-merger \
   --kubevirt-consoleplugin-image-name="${KUBEVIRT_CONSOLE_PLUGIN_IMAGE}" \
   --kubevirt-consoleproxy-image-name="${KUBEVIRT_CONSOLE_PROXY_IMAGE}" \
-  --cli-downloads-image-name="${HCO_DOWNLOADS_IMAGE}" > "${CSV_DIR}/${OPERATOR_NAME}.v${CSV_VERSION}.${CSV_EXT}"
+  --cli-downloads-image-name="${HCO_DOWNLOADS_IMAGE}" > "${CSV_DIR}/${OPERATOR_NAME}.v${CSV_VERSION}.${CSV_EXT}" \
+  --primary-udn-binding-image-name="${PRIMARY_USER_DEFINED_NETWORK_BINDING_IMAGE}"
```
`>` to a file should be the after all the params i think

RamLavi replied:
good catch! thanks!

oshoval commented on tests/func-tests/defaults_test.go:
I think we miss an entry on "Check that featureGates defaults are behaving as expected"

RamLavi replied:
@oshoval added the missing entry.

@nunnatsa `check-defaults.sh` is not checked right? it looks out of sync (many FGs are missing)..

oshoval commented:
sorry that the comments are split, i am less familiar with HCO, and thought on the things in steps
thats it for now

RamLavi commented:
[Change](https://github.com/kubevirt/hyperconverged-cluster-operator/compare/cdc96075d4f2ca648aa6659f8171e10f4a98ac44..2684c4d6d60cc4c6930a76d33333c1f55ef254a9): Rebase

RamLavi commented on tools/csv-merger/csv-merger.go:
```diff
@@ -123,6 +123,7 @@ var (
 	hppoVersion                   = flag.String("hppo-version", "", "HPP operator version")
 	mtqVersion                    = flag.String("mtq-version", "", "MTQ operator version")
 	aaqVersion                    = flag.String("aaq-version", "", "AAQ operator version")
+	primaryUDNImage               = flag.String("primary-udn-binding-image-name", "", "Primary UDN binding image")
```
DONE

RamLavi commented:
[Change](https://github.com/kubevirt/hyperconverged-cluster-operator/compare/2684c4d6d60cc4c6930a76d33333c1f55ef254a9..c444e924ac88448fbf7b19a046c7cca843a8aa08): address @oshoval review

RamLavi commented:
[Change](https://github.com/kubevirt/hyperconverged-cluster-operator/compare/2a1fea056c2e8d1b3a5992a6ed47df5c262bcf23..272863126d5cb80401f20481eeb762c46377132a): Changed so that setting the new env var is a must in order to Init HCO

RamLavi commented:
[Change](https://github.com/kubevirt/hyperconverged-cluster-operator/pull/3034): Fix issue found by linter

RamLavi commented:
@nunnatsa there are a lot of lanes failing - could this be because of the boot condition I added? I don't see the error there though.. Are these lanes usually like this?

RamLavi commented:
[Change](https://github.com/kubevirt/hyperconverged-cluster-operator/compare/aa4b2672afad1a7d547c836272fbf2ea22c00250..66bbd5dfcad45615c22670d890fa89525c8c18c7): Moved the env var check into HCO main function

RamLavi commented:
/hold
I may need to change the NAD's name

RamLavi commented:
[Change](https://github.com/kubevirt/hyperconverged-cluster-operator/compare/66bbd5dfcad45615c22670d890fa89525c8c18c7..7a1a10a331f324f3ac11ab64943688a0fca2bb2c): Changed NAD's name (referenced in the binding's `NetworkAttachmentDefinition` field following review on https://github.com/kubevirt/cluster-network-addons-operator/pull/1841.

nunnatsa commented:
/retest

oshoval commented on deploy/operator.yaml:
```diff
@@ -72,6 +72,8 @@ spec:
         - name: KV_CONSOLE_PLUGIN_IMAGE
         - name: KV_CONSOLE_PROXY_IMAGE
+        - name: PRIMARY_UDN_SIDECAR_IMAGE
+          value: quay.io/kubevirt/network-passt-binding@sha256:3fb5f9fb87a9ab1f06bed2b2358ad862ef66bc47dc73044f1076a3eb9a4c55ff
```
would U/S know to bump this once there is a new kubevirt release ?

nunnatsa commented:
/override-bot

nunnatsa commented:
> /hold I may need to change the NAD's name

the name was changed. Do you wan't to unhold?

RamLavi commented:
> > /hold I may need to change the NAD's name
> 
> the name was changed. Do you wan't to unhold?

I am waiting to confirm the solution manually, then will unhold

RamLavi commented:
[Change](https://github.com/kubevirt/hyperconverged-cluster-operator/compare/040dbc96f4ffd1d5243d055d84701d9cf84ef772..44ecfa73f960c3715be6bbbdc9892d051b0dba79): fixed a parameter in unit tests

RamLavi commented:
[Change](https://github.com/kubevirt/hyperconverged-cluster-operator/compare/44ecfa73f960c3715be6bbbdc9892d051b0dba79..508d61bc0c20514fd493f256b9fad89b6e3f3c5c): Rebase

RamLavi commented on cmd/hyperconverged-cluster-operator/main.go:
```diff
@@ -187,6 +187,9 @@ func main() {
 	cmdHelper.ExitOnError(err, "failed to setup metrics: %v")
 
+	err = hcoutil.CheckPrimaryUDNImageEnvExists()
+	cmdHelper.ExitOnError(err, "failed to retrieve primary UDN env var")
```
`CheckPrimaryUDNImageEnvExists()`  returns the error with the env var name:
```
return fmt.Errorf("%s env var not found", PrimaryUDNImageEnvV)
```

Do you think the message part of the error needs to have it as well?

nunnatsa replied:
cool

RamLavi commented:
/hold cancel
I think it's safe to move forward with this.
/cc @maiqueb

nunnatsa reviewed (approved):
/approve

nunnatsa commented:
/retest
